File: nta/engine/Spec.cpp
/* ---------------------------------------------------------------------
 * Numenta Platform for Intelligent Computing (NuPIC)
 * Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
 * with Numenta, Inc., for a separate license for this software code, the
 * following terms and conditions apply:
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 *
 * http://numenta.org/licenses/
 * ---------------------------------------------------------------------
 */

/** @file 
Implementation of Spec API
*/

#include <nta/engine/Spec.hpp>
#include <nta/utils/Log.hpp>
#include <nta/types/BasicType.hpp>

namespace nta
{




Spec::Spec() : singleNodeOnly(false), description("") 
{
}


std::string Spec::getDefaultInputName() const
{
  if (inputs.getCount() == 0)
    return "";
  if (inputs.getCount() == 1)
    return inputs.getByIndex(0).first;

  // search for default input, but detect multple defaults
  bool found = false;
  std::string name;

  for (size_t i = 0; i < inputs.getCount(); ++i)
  {
    const std::pair<std::string, InputSpec> & p = inputs.getByIndex(i);
    if (p.second.isDefaultInput)
    {
      NTA_CHECK(!found) << "Internal error -- multiply-defined default inputs in Spec";
      found = true;
      name = p.first;
    }
  }
  NTA_CHECK(found) << "Internal error -- multiple inputs in Spec but no default";
  return name;
}

std::string Spec::getDefaultOutputName() const
{
  if (outputs.getCount() == 0)
    return "";
  if (outputs.getCount() == 1)
    return outputs.getByIndex(0).first;

  // search for default output, but detect multple defaults
  bool found = false;
  std::string name;

  for (size_t i = 0; i < outputs.getCount(); ++i)
  {
    const std::pair<std::string, OutputSpec> & p = outputs.getByIndex(i);
    if (p.second.isDefaultOutput)
    {
      NTA_CHECK(!found) << "Internal error -- multiply-defined default outputs in Spec";
      found = true;
      name = p.first;
    }
  }
  NTA_CHECK(found) << "Internal error -- multiple outputs in Spec but no default";
  return name;
}


InputSpec::InputSpec(const std::string& description, 
                     NTA_BasicType  dataType, 
                     UInt32 count,
                     bool required, 
                     bool regionLevel, 
                     bool isDefaultInput,
                     bool requireSplitterMap) :
  description(description), 
  dataType(dataType), 
  count(count), 
  required(required), 
  regionLevel(regionLevel), 
  isDefaultInput(isDefaultInput), 
  requireSplitterMap(requireSplitterMap)
{ 
}

OutputSpec::OutputSpec(const std::string& description, 
                       NTA_BasicType dataType, 
                       size_t count,
                       bool regionLevel, 
                       bool isDefaultOutput) :
  description(description), dataType(dataType), count(count), 
  regionLevel(regionLevel), isDefaultOutput(isDefaultOutput)
{
}


CommandSpec::CommandSpec(const std::string& description) :
  description(description)
{
}


ParameterSpec::ParameterSpec(const std::string& description,
                             NTA_BasicType dataType, 
                             size_t count, 
                             const std::string& constraints, 
                             const std::string& defaultValue, 
                             AccessMode accessMode) :
  description(description), dataType(dataType), count(count),
  constraints(constraints), defaultValue(defaultValue), 
  accessMode(accessMode)
{
  // Parameter of type byte is not supported;
  // Strings are specified as type byte, length = 0
  if (dataType == NTA_BasicType_Byte && count > 0)
    NTA_THROW << "Parameters of type 'byte' are not supported";
}



std::string Spec::toString() const
{
  // TODO -- minimal information here; fill out with the rest of 
  // the parameter spec information
  std::stringstream ss;
  ss << "Spec:" << "\n";
  ss << "Description:" << "\n" 
     << this->description << "\n" << "\n";

  ss << "Parameters:" << "\n";
  for (size_t i = 0; i < parameters.getCount(); ++i)
  {
    const std::pair<std::string, ParameterSpec>& item = parameters.getByIndex(i);
    ss << "  " << item.first << "\n"
       << "     description: " << item.second.description << "\n"
       << "     type: " << BasicType::getName(item.second.dataType) << "\n"
       << "     count: " << item.second.count << "\n"; 
  }

  ss << "Inputs:" << "\n";
  for (size_t i = 0; i < inputs.getCount(); ++i)
  {
    ss << "  " << inputs.getByIndex(i).first << "\n";
  }

  ss << "Outputs:" << "\n";
  for (size_t i = 0; i < outputs.getCount(); ++i)
  {
    ss << "  " << outputs.getByIndex(i).first << "\n";
  }

  ss << "Commands:" << "\n";
  for (size_t i = 0; i < commands.getCount(); ++i)
  {
    ss << "  " << commands.getByIndex(i).first << "\n";
  }
  
  return ss.str();
}


}

File: src/nupic/regions/Spec.py

# ----------------------------------------------------------------------
# Numenta Platform for Intelligent Computing (NuPIC)
# Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
# with Numenta, Inc., for a separate license for this software code, the
# following terms and conditions apply:
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Affero Public License for more details.
#
# You should have received a copy of the GNU Affero Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#
# http://numenta.org/licenses/
# ----------------------------------------------------------------------

"""
## @file
This file defines node spec item classes like InputSpec, OutputSpec,
ParameterSpec and CommandSpec as well as the Spec class itself.
These classes correspond very closely to the NuPIC 2 C++ Spec.

The data type of inputs outputs and parameters must be basic Python types. This
is different from nuPIC 1 that allowed for arbitrary Python values as objects.

The Spec class provides a toDict() method that converts itself a dict of
dicts. All the keys must be regular strings for simplicity (no Unicode strings).

Each item class provides a constructor (__init__() method) with some defaults
and an invariant() method that verifies the validity of initialized object.
"""

dataTypes = ('int', 'uint', 'bool', 'str', 'float', 'Handle')
dataTypesToPyTypes = {
                      'int': int,
                      'uint': int,
                      'bool': bool,
                      'str': str,
                      'float': float,
                      'Handle': object,
                     }

class InputSpec(object):
  def __init__(self,
               description='',
               dataType=None,
               count=1,
               required=False,
               regionLevel=False,
               isDefaultInput=False,
               requireSplitterMap=True):
    self.description=description
    self.dataType=dataType
    self.count=count
    self.required=required
    self.regionLevel=regionLevel
    self.isDefaultInput=isDefaultInput
    self.requireSplitterMap=requireSplitterMap
    self.invariant()

  def invariant(self):
    assert isinstance(self.description, str)
    assert isinstance(self.dataType, str)
    assert self.dataType in dataTypes
    assert isinstance(self.count, int)
    assert self.count >= 0
    assert isinstance(self.required, bool)
    assert isinstance(self.regionLevel, bool)
    assert isinstance(self.isDefaultInput, bool)
    assert isinstance(self.requireSplitterMap, bool)

class OutputSpec(object):
  def __init__(self,
               description='',
               dataType=None,
               count=1,
               regionLevel=False,
               isDefaultOutput=False):
    self.description=description
    self.dataType=dataType
    self.count=count
    self.regionLevel=regionLevel
    self.isDefaultOutput=isDefaultOutput
    self.invariant()

  def invariant(self):
    assert isinstance(self.description, str)
    assert isinstance(self.dataType, str)
    assert self.dataType in dataTypes
    assert isinstance(self.count, int)
    assert self.count >= 0
    assert isinstance(self.regionLevel, bool)
    assert isinstance(self.isDefaultOutput, bool)

class ParameterSpec(object):
  accessModes = ('Create', 'Read', 'ReadWrite')
  def __init__(self,
               description='',
               dataType=None,
               count=1,
               constraints='',
               defaultValue=None,
               accessMode=None):
    self.description=description
    self.dataType=dataType
    # String object can't have fixed length in the parameter spec
    if dataType == 'str':
      count = 0
    self.count=count
    self.constraints=constraints
    self.defaultValue=defaultValue
    self.accessMode=accessMode
    self.invariant()

  def invariant(self):
    assert isinstance(self.description, str)
    assert isinstance(self.dataType, str)
    assert self.dataType in dataTypes
    assert isinstance(self.count, int)
    assert self.count >= 0
    assert isinstance(self.constraints, str)
    # Verify that default value is specified only for 'Create' parameters
    if self.defaultValue is not None:
      assert self.accessMode == 'Create'
      assert isinstance(self.defaultValue, dataTypesToPyTypes[self.dataType])
    else:
      assert self.accessMode in ParameterSpec.accessModes, \
             'Bad access node: ' + self.accessMode

class CommandSpec(object):
  def __init__(self, description):
    self.description = description

  def invariant(self):
    assert isinstance(self.description, str)

class Spec(object):
  def __init__(self, description, singleNodeOnly):
    self.description = description
    self.singleNodeOnly = singleNodeOnly
    self.inputs = {}
    self.outputs = {}
    self.parameters = {}
    self.commands = {}

  def invariant(self):
    """Verify the validity of the node spec object

    The type of each sub-object is verified and then
    the validity of each node spec item is verified by calling
    it invariant() method. It also makes sure that there is at most
    one default input and one default output.
    """
    # Verify the description and singleNodeOnly attributes
    assert isinstance(self.description, str)
    assert isinstance(self.singleNodeOnly, bool)

    # Make sure that all items dicts are really dicts
    assert isinstance(self.inputs, dict)
    assert isinstance(self.outputs, dict)
    assert isinstance(self.parameters, dict)
    assert isinstance(self.commands, dict)

    # Verify all item dicts
    hasDefaultInput = False
    for k, v in self.inputs.items():
      assert isinstance(k, str)
      assert isinstance(v, InputSpec)
      v.invariant()
      if v.isDefaultInput:
        assert not hasDefaultInput
        hasDefaultInput = True


    hasDefaultOutput = False
    for k, v in self.outputs.items():
      assert isinstance(k, str)
      assert isinstance(v, OutputSpec)
      v.invariant()
      if v.isDefaultOutput:
        assert not hasDefaultOutput
        hasDefaultOutput = True

    for k, v in self.parameters.items():
      assert isinstance(k, str)
      assert isinstance(v, ParameterSpec)
      v.invariant()

    for k, v in self.commands.items():
      assert isinstance(k, str)
      assert isinstance(v, CommandSpec)
      v.invariant()

  def toDict(self):
    """Convert the information of the node spec to a plain dict of basic types

    The description and singleNodeOnly attributes are placed directly in
    the result dicts. The inputs, outputs, parameters and commands dicts
    contain Spec item objects (InputSpec, OutputSpec, etc). Each such object
    is converted also to a plain dict using the internal items2dict() function
    (see bellow).
    """

    def items2dict(items):
      """Convert a dict of node spec items to a plain dict

      Each node spec item object will be converted to a dict of its
      attributes. The entire items dict will become a dict of dicts (same keys).
      """
      d = {}
      for k, v in items.items():
        d[k] = v.__dict__

      return d

    self.invariant()
    return dict(description=self.description,
                singleNodeOnly=self.singleNodeOnly,
                inputs=items2dict(self.inputs),
                outputs=items2dict(self.outputs),
                parameters=items2dict(self.parameters),
                commands=items2dict(self.commands))
